### ISSUE TITLE
Create Press Release RSS Feed

### ISSUE THREAD

**Issue by nicolecrom** [NCIOCPL/cgov-digital-platform#298]
This story is to create the RSS Feed for [Press Release](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Content-Type-Press-Release) #155.

* RSS Feed should allow "special characters" (ampersand, less-than, or greater-than) in the RSS feeds.
* All automated RSS feeds are updated every hour (at 20 minutes after the hour), and republished every hour between the hours of 08:00 to 20:00
* On an RSS Feed, the following information from each content item in the RSS feed should appear:
   * The Page Title of each content item
   * The list description of each content item 
      * If no list description exists, then use the meta-description
   * Thumbnail image associated with each content item

**Comment by bryanpizzillo** [comment]
I removed the notes about:
* There should be 4 different RSS display scenarios:
  * Description with image
  * Description with no image
  * No description with image
  * No description with no image

In the description. This only applies to the "create your own RSS feed" content type on current cancer.gov. For press releases, we are handling the RSS feed in a very different way, so it will follow the rules above. (FYI, the press release RSS feed uses the Description with image display...)

**Comment by bgnih** [comment]
Additional Questions:
- How many items should be included in the RSS feed?
- Should the Spanish versions be provided as an RSS feed?
- What URLs should the RSS feeds be at?   /rss/press-release and /espanol/rss/noticias/comunicados-de-prensa ?
- The Drupal RSS feed is generated on-demand.  Does it need to be pre-generated via a cron job instead?

**Comment by nicolecrom** [comment]
Based on the old RTM and what I found in percussion here are my answers but please confirm @lburack and @VictoriaSunNIH 

* Posts - 10
* Separate Spanish Feed
* English link is https://www.cancer.gov/PublishedContent/RSS/syndication/rss/ncinewsreleases.rss
* Spanish Link is https://www.cancer.gov/PublishedContent/RSS/syndication/rss/nci-noticias.rss
* Publishing: 
  * The system shall republish all automated RSS feeds every hour (at 20 minutes after the hour) (The business reason behind this is to allow new press releases to appear in the autoRSS, which is used for GovDelivery. When a new press release goes live on Cancer.Gov, we should republish https://www.cancer.gov/PublishedContent/rss/syndication/rss/ncinewsreleases.rss)
  * The system shall republish every hour between the hours of 08:00 to 20:00

**Comment by bgnih** [comment]
@bryanpizzillo There is an open issue on core Feeds where it does not show languages properly: [RSS views do not show nodes in the correct language](https://www.drupal.org/project/drupal/issues/2891923])

**Comment by bgnih** [comment]
New EN RSS: http://www.devbox/PublishedContent/RSS/syndication/rss/ncinewsreleases.rss
New ES RSS: http://www.devbox/PublishedContent/RSS/syndication/rss/nci-noticias.rss

Prod EN RSS: https://www.cancer.gov/PublishedContent/RSS/syndication/rss/ncinewsreleases.rss
Prod ES RSS: https://www.cancer.gov/PublishedContent/RSS/syndication/rss/nci-noticias.rss

Notes: 

- I got the Spanish version working changing the feed from Content to Fields-based format. 
- View custom text field rewrites display the appropriate lead/promo image and list/meta descriptions.
- The core RSS feed doesn't support `<enclosure>` tags on items, so I added the image to the `<description>` tag, which is the alternate way I've seen it implemented.  If you want me to work more on adding the enclosure tags feature let me know.  An extra wrinkle will be that the images are in media entities, not image fields.  Views_RSS which supported `<enclosure>` in D7 doesn't have a working D8 version yet.
- The date field in the current prod is not in RFC822 format, the new implementation is.
- The current feed header uses `<language>es-us</language>`, but the D8 feed is stuck on `<language>en</language>` (showing `en`, and different format (no `-us`) )

**Comment by bgnih** [comment]
There is a problem where when adding the multiple RSS feeds within a single view, when you change the title on a display, it causes an problem with the view's config/schema.  

As a result of this, it causes Functional tests to fail (I think when it's setting up the profile) with an error like:

```
views.view.test:display.feed_1.display_options.defaults.sitename_title missing schema
```
before the actual asserts even start.

The issue is visible in the UI when using the [Configuration Inspector](https://www.drupal.org/project/config_inspector) module.  

I found an issue [Missing schema for display_options.defaults.sitename_title](https://www.drupal.org/project/drupal/issues/2952990) with similar behavior, but the patch on the ticket did not solve the resolve the issue.

As a workaround I have created two separate views, each containing one of the feeds.

**Comment by bgnih** [comment]
### RSS `<enclosure>` Research:

- Core `Views` `Feed` display
  - `Fields` row style (currently being used)
    - Pro: Allows field rewrites to conditionally display Promo/Lead Image and List/Meta Descriptions
    - Con: Does not support Enclosures field
    - Possible Solution: 
      - Create a core patch that adds the Enclosures field that supports both File and Media fields types.
  - `Content` row style:  
    - Pro: display supports File field enclosures using the RSS Enclosures field formatter in an RSS display mode ([Added here](https://www.drupal.org/project/drupal/issues/666412) when Media wasn't in core)
    - Con: Does not support Media field enclosures.
    - Con: Can not rewrite fields to conditionally display Promo/Lead Image and List/Meta Descriptions
  - Possible Solutions: 
    - Core patch to extend RSS Enclosures field formatter to support Media fields
    - Custom module to add Media field Enclosure field formatter
    - Caveat: Would still need way of doing field rewrites to conditionally display Promo/Lead Image and List/Meta Descriptions
- [Views RSS](https://www.drupal.org/project/views_rss) contrib module
  - No stable D8 version
  - Does not support Media fields for enclosures either

Note 1: The `/admin/config/services/rss-publishing` link does not actually do anything, and has a 3yr-old issue to remove it: [Remove fields that do nothing from the "RSS publishing" settings form](https://www.drupal.org/project/drupal/issues/2409413)

Note 2: There is more screwed-up-ness on RSS as detailed in [Views RSS view mode settings are completely broken](https://www.drupal.org/project/drupal/issues/2601030)

Relevant system code: 
```
RSS TWIG TEMPLATE:
docroot/core/modules/views/templates/views-view-rss.html.twig:
<?xml version="1.0" encoding="utf-8" ?>
<rss version="2.0" xml:base="{{ link }}"{{ namespaces }}>
  <channel>
    <title>{{ title }}</title>
    <link>{{ link }}</link>
    <description>{{ description }}</description>
    <language>{{ langcode }}</language>
    {{ channel_elements }}
    {{ items }}
  </channel>
</rss>


RSS ROW TWIG TEMPLATE:
docroot/core/modules/views/templates/views-view-row-rss.html.twig:
<item>
  <title>{{ title }}</title>
  <link>{{ link }}</link>
  <description>{{ description }}</description>
  {% for item in item_elements -%}
    <{{ item.key }}{{ item.attributes -}}
    {% if item.value -%}
      >{{ item.value }}</{{ item.key }}>
    {% else -%}
      {{ ' />' }}
    {% endif %}
  {%- endfor %}
</item>


RSS FORMATTER FOR FILE TYPES:
docroot/core/modules/file/src/Plugin/Field/FieldFormatter/RSSEnclosureFormatter.php:
/**
* Plugin implementation of the 'file_rss_enclosure' formatter.
*
* @FieldFormatter(
*   id = "file_rss_enclosure",
*   label = @Translation("RSS enclosure"),
*   field_types = {
*     "file",
*     "media"
*   }
* )
*/
class RSSEnclosureFormatter extends FileFormatterBase {

  /**
   * {@inheritdoc}
   */
  public function viewElements(FieldItemListInterface $items, $langcode) {
    $entity = $items->getEntity();
    // Add the first file as an enclosure to the RSS item. RSS allows only one
    // enclosure per item. See: http://wikipedia.org/wiki/RSS_enclosure
    foreach ($this->getEntitiesToView($items, $langcode) as $delta => $file) {
      /** @var \Drupal\file\FileInterface $file */
      $entity->rss_elements[] = [
        'key' => 'enclosure',
        'attributes' => [
          // In RSS feeds, it is necessary to use absolute URLs. The 'url.site'
          // cache context is already associated with RSS feed responses, so it
          // does not need to be specified here.
          'url' => $file->createFileUrl(FALSE),
          'length' => $file->getSize(),
          'type' => $file->getMimeType(),
        ],
      ];
    }
    return [];
  }

}

THEME RSS HEADER:
docroot/core/modules/views/views.theme.inc:

/**
 * Prepares variables for RSS feed templates.
 *
 * Default template: views-view-rss.html.twig.
 *
 * @param array $variables
 *   An associative array containing:
 *   - view: A ViewExecutable object.
 *   - rows: The raw row data.
 */
function template_preprocess_views_view_rss(&$variables) {
  $view = $variables['view'];
  $items = $variables['rows'];
  $style = $view->style_plugin;

  $config = \Drupal::config('system.site');

  // The RSS 2.0 "spec" doesn't indicate HTML can be used in the description.
  // We strip all HTML tags, but need to prevent double encoding from properly
  // escaped source data (such as &amp becoming &amp;amp;).
  $variables['description'] = Html::decodeEntities(strip_tags($style->getDescription()));

  if ($view->display_handler->getOption('sitename_title')) {
    $title = $config->get('name');
    if ($slogan = $config->get('slogan')) {
      $title .= ' - ' . $slogan;
    }
  }
  else {
    $title = $view->getTitle();
  }
  $variables['title'] = $title;

  // Figure out which display which has a path we're using for this feed. If
  // there isn't one, use the global $base_url
  $link_display_id = $view->display_handler->getLinkDisplay();
  /** @var \Drupal\views\Plugin\views\display\DisplayPluginBase $display */
  if ($link_display_id && ($display = $view->displayHandlers->get($link_display_id)) && $display->isEnabled()) {
    $url = $view->getUrl(NULL, $link_display_id);
  }

  /** @var \Drupal\Core\Url $url */
  if (!empty($url)) {
    $url_options = ['absolute' => TRUE];
    if (!empty($view->exposed_raw_input)) {
      $url_options['query'] = $view->exposed_raw_input;
    }

    // Compare the link to the default home page; if it's the default home page,
    // just use $base_url.
    $url_string = $url->setOptions($url_options)->toString();
    if ($url_string === Url::fromUserInput($config->get('page.front'))->toString()) {
      $url_string = Url::fromRoute('<front>')->setAbsolute()->toString();
    }

    $variables['link'] = $url_string;
  }

  $variables['langcode'] = \Drupal::languageManager()->getCurrentLanguage()->getId();
  $variables['namespaces'] = new Attribute($style->namespaces);
  $variables['items'] = $items;
  $variables['channel_elements'] = \Drupal::service('renderer')->render($style->channel_elements);

  // During live preview we don't want to output the header since the contents
  // of the feed are being displayed inside a normal HTML page.
  if (empty($variables['view']->live_preview)) {
    $variables['view']->getResponse()->headers->set('Content-Type', 'application/rss+xml; charset=utf-8');
  }
}

THEME RSS ROWS:
docroot/core/modules/views/views.theme.inc:

/**
 * Prepares variables for views RSS item templates.
 *
 * Default template: views-view-row-rss.html.twig.
 *
 * @param array $variables
 *   An associative array containing:
 *   - row: The raw results rows.
 */
function template_preprocess_views_view_row_rss(&$variables) {
  $item = $variables['row'];
  $variables['title'] = $item->title;
  $variables['link'] = $item->link;

  // The description is the only place where we should find HTML.
  // @see https://validator.w3.org/feed/docs/rss2.html#hrelementsOfLtitemgt
  // If we have a render array, render it here and pass the result to the
  // template, letting Twig autoescape it.
  if (isset($item->description) && is_array($item->description)) {
    $variables['description'] = (string) \Drupal::service('renderer')->render($item->description);
  }

  $variables['item_elements'] = [];
  foreach ($item->elements as $element) {
    if (isset($element['attributes']) && is_array($element['attributes'])) {
      $element['attributes'] = new Attribute($element['attributes']);
    }
    $variables['item_elements'][] = $element;
  }
}
```

**Comment by bryanpizzillo** [comment]
@bgnih I am having a hard time following what the root issue is. 
1. Do you have the ability to modify the RSS twig templates, specifically views-view-row-rss.html.twig?
2. Can the view return all the fields we would need to render? (Such that you add <enclosure>...</enclosure> to the twig)

**Comment by bgnih** [comment]
@bryanpizzillo I was more asking which approach I should take to resolve it.   I would personally favor the core patch so that it will be fixed for all sites in the future and not have to re-hack every site to get it to work, but wanted to confirm that that would be okay first.

**Comment by bgnih** [comment]
# Cgov RSS Feed Approaches
## Drupal Core Functionality

Drupal 8 core has the **Feeds** view display which can display using the *Content* format or the *Fields* format.
- If using the **Content** format it will attempt to use the RSS display mode.
  - [Enclosures](https://en.wikipedia.org/wiki/RSS_enclosure) must be of type *File* and should use the *RSS Formatter* formatter to display it properly.  There is no support for Image or Media fields to be used as enclosures.  [Added in 8.3](https://www.drupal.org/project/drupal/issues/666412)
- If using the **Fields** format you need to specify how to map the available fields into the RSS fields (Description, path, etc)
  - There is no support for Enclosures using the *Fields* format.   
  - There is a [patch to add Enclosure support](https://www.drupal.org/project/drupal/issues/2511878) but it only supports File, not Image or Media.
- None of these options support the conditional image logic needed on Cgov.
- Supports adding the view as an attachement to another view, which we probably don't need
- Adds the RSS mime type via `$variables['view']->getResponse()->headers->set('Content-Type', 'application/rss+xml; charset=utf-8');` (if not in Preview mode)

## Approaches: Getting correct image data

### Determine correct image file and mode to display
- Display Promo Image or Main Image (`cgov_image` field)
  - Choose appropriate display mode [Image Crop: Article, Image Crop, Thumbnail, etc]
  - If there is an override image for that display mode [Thumbnail, Featured, Panoramic, Social Media] display that instead (`media` field)
  - Get the File object for the needed image
    - Apply the proper image style
      - Generate the image in the filesystem
        - Return the absolute URL, Mime Type and File Size (needed for the <enclosure> tag)

### Return data via a Service to provide enclosure information
Create a service that will return the needed image info and metadata.
Caveats: Service would need to generate the appropriate image in the proper style to the filesystem in order to get the file size.  
```
$imginfo = cgov_image::get_image_crop_info($eid, 'featured')
$url = $imginfo.absolute_url;
$mime_type = $imaginfo.mime_type;
$length = $imginfo.length;
...
return: <enclosure src='absolute_url' type='mime_type' length='$length'>
```

### Get data via Twig function
Put display logic in Twig layer, eg `{% set thumbnail = get_thumbnail_image(id) %}`
Can be called directly from twig templates (or anywhere that takes Twig, such as a views field rewrite)
Can also just be a wrapper around `cgov_image::Get_image_crop_info()` (or whatever call is used)

### Store data in Computed Field 
Store intended Thumbnail (or other) image to be used as a computed field (db storage) on the node that can be accessed without further processing.  Caveat: Need to create wrapper code for views to be able to use computed field



## Approaches: Output RSS feed

### Patch Core RSS Feeds with Fields format
Would require patching core to [allow Enclosure fields](https://www.drupal.org/project/drupal/issues/2511878) to be specified.   No node information is currently passed to the RssRow template, so we'd have to add it to the preprocess function

Could use a Twig function {{ cgov_image(nid, 'thumbnail') }} or similar in the field rewrite that would return a file object that could have the core RSS_Formatter applied to it.   Might have issues with field data returned from twig function in views via a rewrite option.

### Use existing Core RSS Feeds with Fields format

### Core RSS Feeds with Content format and RSS display mode
Would need to add a preprocess function to attach row/node/image data so that the template can create an <enclosure> tag for each row.  (No row data is currently sent to the template except for the chosen `item_elements` variables displayed in the feed in the `views-view-row-rss.html.twig` template.

### Create a generic page view using Unformatted List view and update view templates for RSS layout
Use a standard Unformatted List view of Content and use the RSS view mode and custom Twig template files in the theme to override the output.  Would need to be applied in a parent theme or it wouldn't work across multiple themes.  Would need to add mime-type header to output.

**Comment by bryanpizzillo** [comment]
The Approach to Take:
* Add the RSS Display mode to the Press Releases
   * Add the required fields, including the field_image_article and field_image_promotional each using the Rendered Entity formatter with the Crop: Thumbnail display style.
* Use the Content format for the feed
   * Please make the machine name for the display for the feed to be `pr_feed`
* Create RSS templates for the Press Release View's Row (`views-view-row-rss--press-releases.html.twig`)
   * This will allow you to add your own enclosure

* While Drupal 8 has no OOB support for making the enclosures for you, you can override the template and make your own markup. (With logic for figuring out which image to use)

Given this is possible to just override the template. You can make a twig extension that when given an image field, it can tell you how big the file is and what the mime type is.  Then call it from views-view-row-rss--press-releases.html.twig.

**Comment by bryanpizzillo** [comment]
## `cgov_image` media type rendering logic for the `Image Crop: Thumbnail` display mode

### Setup for the `cgov_image`
The `Image Crop: Thumbnail` Display Mode includes both:
* `field_media_image` using the Thumbnail (`cgov_thumbnail`) Image Style
* `field_override_img_thumbnail` using the Thumbnail (`cgov_thumbnail`) image style

### Render Array Manipulation
A `cgov_image_media_view()` hook then processes the `$build` structure for the `Image Crop: Thumbnail` display mode.
* If the `override_img_thumbnail` exists, we remove `field_media_image` from the `$build` structure
* if no `override_img_thumbnail`, remove `override_img_thumbnail` from the `$build` Structure
  * If the thumbnail crop, using the Image Widget Crop tool has NOT been applied to `field_media_image` then we remove `field_media_image` from the `$build` structure.

### Rendering Template
This would allow the `media--cgov-image--image-crop-thumbnail.html.twig` to only have to contain:
```
{{ content }}
``` 

In fact, we only have a `media--cgov-image.html.twig` that contains a large comment about what is going on and the `{{content}}` code block.

**Comment by bgnih** [comment]
When rendering with the Content format, all fields in the display are rendered as the contents of the `<description>` tag, and are not individually accessible in the row template.  Only the NID is passed in.  I'll try passing this NID to a twig template and determining the proper image in the function there.
![image.png](https://images.zenhubusercontent.com/5c0fd0a81a65780cf410b298/2b87d3f2-8315-4788-bc9d-929657597444)

**Comment by bgnih** [comment]
Yet another RSS Feeds issue: [Views RSS display always outputs ‘submitted’ by and ‘authored on date’ inside the description tag](https://www.drupal.org/project/drupal/issues/2946812)

Despite not being in the RSS display at all, the RSS feed (Content format) is adding Author/Changed Date/Title to all `<item>.<description>` tags, eg:

```
Anonymous (not verified) Sun, 03/17/2019 - 17:31 AI approach outperformed human experts in identifying cervical precancer
```

**Comment by bgnih** [comment]
Another Feeds issue: [How to get feed title in correct language?](https://www.drupal.org/project/drupal/issues/2816447): When viewing feed in spanish (requiring /espanol/ in front of URL) node titles are in English, though Body is in Spanish

**Comment by bgnih** [comment]
Ended up reworking this to use the Fields display which did not suffer from the "added fields at the bottom" and "title in wrong language" core bugs.   Used a hack in the approach to pass the NID in through the GUID variable (since the row template does not expose the node or NID as variables, and this was an easy way to pass it in to the generally-broken RSS Feed view display without additional code.   The template then uses the link value for the GUID.  The NID is required to build the enclosure tag and switch between List/Page description values via Twig functions.

**Comment by bgnih** [comment]
Notes from Pull Request #1095:  

Fixes #298.   

EN: /PublishedContent/RSS/syndication/rss/ncinewsreleases.rss
ES: /espanol/PublishedContent/RSS/syndication/rss/nci-noticias.rss 

`<description>` uses List Description (if present) or Meta Description.
`<enclosure>` uses last image of:
1. Lead Image
1. Lead Image thumbnail override
1. Promo Image
1. Promo Image thumbnail override

and renders using the `cgov_thumbnail` style. 

[Known issues](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Drupal-Issues-Watchlist#rss-feeds):   
1. The date format is different in this version than on current PROD.   This version is correct, PROD is not using the proper date format, per the [RSS spec](http://www.rssboard.org/rss-specification).
1. This version uses "escaped encoding" instead of the "CDATA encoding" that current PROD version uses.   Either is fine per the RSS spec.
1. This version uses `en` and `es` as the `channel.language` value.  The current PROD uses `en-us` and `es-us`.  Both are permitted by the RSS spec.
1. The Spanish page will needs a redirect from `/PublishedContent/RSS/syndication/rss/nci-noticias.rss` to `/espanol/PublishedContent/RSS/syndication/rss/nci-noticias.rss` to support existing feeds.  (node renders in EN without the /espanol/ prefix).  This can be done with an Akamai reroute, or other solution.

**Comment by bgnih** [comment]
More Core issues with translated node paths in views not working properly, used twig function to get proper data, eg:
   * https://www.drupal.org/project/drupal/issues/2877994
   * https://www.drupal.org/project/drupal/issues/2957917
   * https://www.drupal.org/project/drupal/issues/2862511
   * https://www.drupal.org/project/drupal/issues/2802311
